src/Compilers/CSharp/Test/Semantic/SourceGeneration/SyntaxAwareGeneratorTests.cs

```diff
@@ -192,21 +193,247 @@ void Function()
+                onExecute: (e) => { Assert.True(false); }
+                );
+
+            GeneratorDriver driver = new CSharpGeneratorDriver(parseOptions, ImmutableArray.Create<ISourceGenerator>(testGenerator), CompilerAnalyzerConfigOptionsProvider.Empty, ImmutableArray<AdditionalText>.Empty);
+            driver = driver.RunFullGeneration(compilation, out var outputCompilation, out var outputDiagnostics);
+            var results = driver.GetRunResult();
+
+            Assert.Empty(results.GeneratedTrees);
+            Assert.Single(results.Diagnostics);
+            Assert.Single(results.Results);
+            Assert.Single(results.Results[0].Diagnostics);
+
+            Assert.NotNull(results.Results[0].Exception);
+            Assert.Equal("Test Exception", results.Results[0].Exception?.Message);
+
+            outputDiagnostics.Verify(
+                Diagnostic(ErrorCode.WRN_GeneratorFailedDuringGeneration).WithArguments("CallbackGenerator").WithLocation(1, 1)
+                );
+        }
+
+        [Fact]
+        public void Syntax_Receiver_Exception_During_Visit()
+        {
+            var source = @"
+class C 
+{
+    int Property { get; set; }
+
+    void Function()
+    {
+        var x = 5;
+        x += 4;
+    }
+}
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            var testGenerator = new CallbackGenerator(
+                onInit: (i) => i.RegisterForSyntaxNotifications(() => new TestSyntaxReceiver(tag: 0, callback: (a) => { if (a is AssignmentExpressionSyntax) throw new Exception("Test Exception"); })),
+                onExecute: (e) => { e.AddSource("test", SourceText.From("public class D{}", Encoding.UTF8)); }
+                );
+
+            GeneratorDriver driver = new CSharpGeneratorDriver(parseOptions, ImmutableArray.Create<ISourceGenerator>(testGenerator), CompilerAnalyzerConfigOptionsProvider.Empty, ImmutableArray<AdditionalText>.Empty);
+            driver = driver.RunFullGeneration(compilation, out var outputCompilation, out var outputDiagnostics);
+            var results = driver.GetRunResult();
+
+            Assert.Empty(results.GeneratedTrees);
+            Assert.Single(results.Diagnostics);
+            Assert.Single(results.Results);
+            Assert.Single(results.Results[0].Diagnostics);
+
+            Assert.NotNull(results.Results[0].Exception);
+            Assert.Equal("Test Exception", results.Results[0].Exception?.Message);
+
+            outputDiagnostics.Verify(
+                Diagnostic(ErrorCode.WRN_GeneratorFailedDuringGeneration).WithArguments("CallbackGenerator").WithLocation(1, 1)
+                );
+        }
+
+        [Fact]
+        public void Syntax_Receiver_Exception_During_Visit_Stops_Visits_On_Other_Trees()
```

Comment:
>Syntax_Receiver_Exception_During_Visit_Stops_Visits_On_Other_Trees [](start = 20, length = 66)

It's not clear to me what is being tested. It looks like there only one syntax tree, so which trees are skipped?